File: apps/api/app/interfaces/api/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class ApiError(BaseModel):
    code: str
    message: str


class RegisterRequest(BaseModel):
    email: EmailStr
    password: str = Field(min_length=12, max_length=128)
    captcha_token: str | None = Field(default=None, max_length=4096)
    site_owner_bootstrap_token: str | None = Field(default=None, max_length=4096)


class LoginRequest(RegisterRequest):
    captcha_token: None = None
    site_owner_bootstrap_token: None = None
    mfa_code: str | None = Field(default=None, max_length=64)


class VerifyEmailRequest(BaseModel):
    token: str


class PasswordResetRequest(BaseModel):
    email: EmailStr
    captcha_token: str | None = Field(default=None, max_length=4096)


class PasswordResetConfirmRequest(BaseModel):
    token: str
    password: str = Field(min_length=12, max_length=128)


class UserView(BaseModel):
    id: str
    email: EmailStr
    is_verified: bool
    account_type: str = "user"
    account_status: str = "active"
    model_config = ConfigDict(from_attributes=True)


class WorkspaceView(BaseModel):
    id: str
    name: str
    model_config = ConfigDict(from_attributes=True)


class MfaRequirementView(BaseModel):
    required: bool = False
    authenticator_app_enabled: bool = False
    passkey_registered: bool = False
    passkey_verified: bool = False
    setup_required: bool = False
    passkey_verification_required: bool = False


class AuthResponse(BaseModel):
    user: UserView
    workspace: WorkspaceView
    workspace_role: str | None = None
    csrf_token: str | None = None
    verification_token: str | None = None
    reset_token: str | None = None
    mfa_requirements: MfaRequirementView = Field(default_factory=MfaRequirementView)
    mfa_setup_required: bool = False
    passkey_verification_required: bool = False


class CaptchaChallengeView(BaseModel):
    required: bool
    provider: Literal["disabled", "turnstile", "challenge"]
    token: str = Field(default="", max_length=4096)
    prompt: str = Field(default="", max_length=120)
    expires_in_seconds: int = Field(default=0, ge=0, le=900)


class MfaStatusView(BaseModel):
    enabled: bool
    required: bool = False


class MfaSetupView(BaseModel):
    enabled: bool
    secret: str
    provisioning_uri: str
    recovery_codes: list[str]


class MfaCodeRequest(BaseModel):
    code: str = Field(min_length=6, max_length=64)


class PasskeyCredentialView(BaseModel):
    id: str
    name: str
    created_at: datetime
    last_used_at: datetime | None = None


class PasskeyStatusView(MfaRequirementView):
    registered: bool = False
    count: int = 0
    credentials: list[PasskeyCredentialView] = Field(default_factory=list)


class PasskeyOptionsView(BaseModel):
    options: dict[str, Any]


class PasskeyRegistrationVerifyRequest(BaseModel):
    credential: dict[str, Any]
    name: str | None = Field(default=None, max_length=120)


class PasskeyAuthenticationVerifyRequest(BaseModel):
    credential: dict[str, Any]


class ProjectCreate(BaseModel):
    workspace_id: str
    title: str = Field(min_length=1, max_length=220)
    description: str = Field(default="", max_length=5000)


class ProjectUpdate(BaseModel):
    title: str = Field(min_length=1, max_length=220)
    description: str = Field(default="", max_length=5000)


class ProjectView(BaseModel):
    id: str
    workspace_id: str
    created_by_user_id: str | None = None
    title: str
    description: str
    model_config = ConfigDict(from_attributes=True)


class ReviewCreate(BaseModel):
    title: str = Field(min_length=1, max_length=220)
    proposal_text: str = Field(min_length=1, max_length=50000)
    mode: Literal["basic", "standard", "in_depth"] = "standard"
    focus_chips: list[str] = Field(default_factory=list, max_length=12)
    external_research: bool = False
    private_research: bool = True
    domain_allowlist: list[str] = Field(default_factory=list, max_length=25)
    domain_blocklist: list[str] = Field(default_factory=list, max_length=25)


class StandaloneReviewCreate(ReviewCreate):
    workspace_id: str


class ReviewUpdate(ReviewCreate):
    pass


class ReviewView(BaseModel):
    id: str
    workspace_id: str
    project_id: str | None
    title: str
    proposal_text: str
    mode: str
    focus_chips: list[str]
    external_research: bool
    private_research: bool
    domain_allowlist: list[str]
    domain_blocklist: list[str]
    model_config = ConfigDict(from_attributes=True)


class PastedTextRequest(BaseModel):
    text: str = Field(min_length=1, max_length=50000)


class WebsiteSourceRequest(BaseModel):
    url: str = Field(min_length=8, max_length=2048)


class RepositorySourceRequest(BaseModel):
    url: str = Field(min_length=8, max_length=2048)


class SourceView(BaseModel):
    id: str
    workspace_id: str
    review_id: str
    filename: str
    content_type: str
    state: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)


class ContextPackCreate(BaseModel):
    workspace_id: str
    name: str = Field(min_length=1, max_length=220)
    agent_key: str = Field(min_length=1, max_length=80)
    markdown: str = Field(min_length=1, max_length=50000)


class ContextPackView(BaseModel):
    id: str
    workspace_id: str
    name: str
    agent_key: str
    markdown: str
    version: int
    model_config = ConfigDict(from_attributes=True)


class ProviderConnectionCreate(BaseModel):
    workspace_id: str
    adapter: str = Field(min_length=1, max_length=80)
    name: str = Field(min_length=1, max_length=160)
    config: dict[str, Any] = Field(default_factory=dict)
    credentials: dict[str, str] = Field(default_factory=dict)


class ProviderConnectionView(BaseModel):
    id: str
    workspace_id: str
    adapter: str
    name: str
    config: dict[str, Any]
    has_credentials: bool


class ModelPreviewRequest(BaseModel):
    workspace_id: str
    adapter: str = Field(min_length=1, max_length=80)
    config: dict[str, Any] = Field(default_factory=dict)
    credentials: dict[str, str] = Field(default_factory=dict)


class ModelPreviewView(BaseModel):
    model_identifier: str
    capabilities: list[str]
    provenance: str
    verified: bool
    probe_result: dict[str, Any] = Field(default_factory=dict)


class ModelCreate(BaseModel):
    workspace_id: str
    provider_connection_id: str
    model_identifier: str = Field(min_length=1, max_length=160)
    capabilities: list[str] = Field(default_factory=list, max_length=20)
    provenance: str = Field(default="manual", max_length=120)
    verified: bool = False


class ModelView(BaseModel):
    id: str
    workspace_id: str
    provider_connection_id: str
    model_identifier: str
    capabilities: list[str]
    provenance: str
    verified: bool
    probe_result: dict[str, Any] = Field(default_factory=dict)
    model_config = ConfigDict(from_attributes=True)


class ProfileCreate(BaseModel):
    workspace_id: str
    name: str = Field(min_length=1, max_length=160)
    agent_key: str = Field(min_length=1, max_length=80)
    model_record_id: str
    explicit_pin: bool = False


class ProfileView(BaseModel):
    id: str
    workspace_id: str
    name: str
    agent_key: str
    model_record_id: str
    explicit_pin: bool
    model_config = ConfigDict(from_attributes=True)


class RunView(BaseModel):
    id: str
    workspace_id: str
    review_id: str
    state: str
    routing_plan: dict[str, Any]
    usage: dict[str, Any]
    model_config = ConfigDict(from_attributes=True)


class UsageLimitsView(BaseModel):
    account_type: str
    tier_name: str
    project_limit: int | None
    projects_used: int
    projects_remaining: int | None
    workflow_total_limit: int | None
    workflows_used: int
    workflows_remaining: int | None
    workflow_weekly_limit: int | None
    workflows_started_this_week: int
    weekly_workflows_remaining: int | None
    resets_at: datetime
    daily_review_run_limit: int | None = None
    runs_started_today: int = 0
    runs_remaining_today: int | None = None


class WorkflowSummaryView(BaseModel):
    id: str
    workspace_id: str
    review_id: str
    review_title: str
    project_id: str | None
    project_title: str
    mode: str
    state: str
    created_at: datetime
    selected_agents: list[str]
    top_risks: list[str]
    finding_count: int
    has_report: bool


class RunEventView(BaseModel):
    id: str
    run_id: str
    state: str
    message: str
    sequence: int
    model_config = ConfigDict(from_attributes=True)


class ReportView(BaseModel):
    id: str
    workspace_id: str
    run_id: str
    data: dict[str, Any]
    model_config = ConfigDict(from_attributes=True)


class ReportComparisonView(BaseModel):
    left_run_id: str
    right_run_id: str
    changed_risks: list[str]
    changed_assumptions: list[str]
    changed_evidence_gaps: list[str]
    changed_recommendations: list[str]


class EvaluationResultView(BaseModel):
    workspace_id: str
    fixture_count: int
    metrics: dict[str, float]
    adversarial_fixtures: list[str]
    live_smoke_tests: str


def source_view(source: Any) -> SourceView:
    return SourceView(
        id=source.id,
        workspace_id=source.workspace_id,
        review_id=source.review_id,
        filename=source.filename,
        content_type=source.content_type,
        state=source.state,
        metadata=source.metadata_json,
        warnings=source.warnings,
    )

File: apps/web/src/api/schemas.ts
import { z } from 'zod';

export { reportComparisonSchema, reportSchema } from './reportSchemas';

export const API_BASE = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000';

export const authSchema = z.object({
  user: z.object({
    id: z.string(),
    email: z.string(),
    is_verified: z.boolean(),
    account_type: z.enum(['owner', 'admin', 'user']).default('user'),
    account_status: z.enum(['active', 'suspended', 'banned', 'deleted']).default('active')
  }),
  workspace: z.object({ id: z.string(), name: z.string() }),
  workspace_role: z.string().nullable().optional(),
  csrf_token: z.string().nullable().optional(),
  verification_token: z.string().nullable().optional(),
  reset_token: z.string().nullable().optional(),
  mfa_requirements: z.object({
    required: z.boolean().default(false),
    authenticator_app_enabled: z.boolean().default(false),
    passkey_registered: z.boolean().default(false),
    passkey_verified: z.boolean().default(false),
    setup_required: z.boolean().default(false),
    passkey_verification_required: z.boolean().default(false)
  }).default({
    required: false,
    authenticator_app_enabled: false,
    passkey_registered: false,
    passkey_verified: false,
    setup_required: false,
    passkey_verification_required: false
  }),
  mfa_setup_required: z.boolean().default(false),
  passkey_verification_required: z.boolean().default(false)
});

export const captchaChallengeSchema = z.object({
  required: z.boolean(),
  provider: z.enum(['disabled', 'turnstile', 'challenge']),
  token: z.string(),
  prompt: z.string(),
  expires_in_seconds: z.number()
});

export const mfaStatusSchema = z.object({
  enabled: z.boolean(),
  required: z.boolean().default(false)
});

export const mfaSetupSchema = z.object({
  enabled: z.boolean(),
  secret: z.string(),
  provisioning_uri: z.string(),
  recovery_codes: z.array(z.string())
});

export const siteUserSchema = z.object({
  id: z.string(),
  email: z.string(),
  is_verified: z.boolean(),
  account_type: z.enum(['owner', 'admin', 'user']),
  account_status: z.enum(['active', 'suspended', 'banned', 'deleted']),
  status_message: z.string().default(''),
  admin_scope: z.enum(['none', 'all', 'selected']).default('none'),
  admin_managed_user_ids: z.array(z.string()).default([]),
  created_at: z.string(),
  last_login_at: z.string().nullable().optional(),
  last_login_ip: z.string().nullable().optional(),
  last_seen_at: z.string().nullable().optional(),
  last_seen_ip: z.string().nullable().optional(),
  run_count: z.number().default(0)
});

export const siteVisitSchema = z.object({
  id: z.string(),
  user_id: z.string().nullable().optional(),
  ip_address: z.string(),
  method: z.string(),
  path: z.string(),
  user_agent: z.string(),
  created_at: z.string()
});

export const projectSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  created_by_user_id: z.string().nullable().optional(),
  title: z.string(),
  description: z.string()
});

export const reviewSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  project_id: z.string().nullable(),
  title: z.string(),
  proposal_text: z.string(),
  mode: z.enum(['basic', 'standard', 'in_depth']),
  focus_chips: z.array(z.string()),
  external_research: z.boolean().default(false),
  private_research: z.boolean().default(true),
  domain_allowlist: z.array(z.string()).default([]),
  domain_blocklist: z.array(z.string()).default([])
});

export const sourceSchema = z.object({
  id: z.string(),
  filename: z.string(),
  content_type: z.string(),
  state: z.enum(['pending', 'ingested', 'failed']),
  metadata: z.record(z.string(), z.unknown()),
  warnings: z.array(z.string())
});

export const contextPackSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  name: z.string(),
  agent_key: z.string(),
  markdown: z.string(),
  version: z.number()
});

export const runSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  review_id: z.string(),
  state: z.string(),
  routing_plan: z.record(z.string(), z.unknown()),
  usage: z.record(z.string(), z.unknown())
});

export const usageLimitsSchema = z.object({
  account_type: z.enum(['owner', 'admin', 'user']).default('user'),
  tier_name: z.string().default('User'),
  project_limit: z.number().nullable().default(null),
  projects_used: z.number().default(0),
  projects_remaining: z.number().nullable().default(null),
  workflow_total_limit: z.number().nullable().default(null),
  workflows_used: z.number().default(0),
  workflows_remaining: z.number().nullable().default(null),
  workflow_weekly_limit: z.number().nullable().default(null),
  workflows_started_this_week: z.number().default(0),
  weekly_workflows_remaining: z.number().nullable().default(null),
  daily_review_run_limit: z.number().nullable().default(null),
  runs_started_today: z.number().default(0),
  runs_remaining_today: z.number().nullable().default(null),
  resets_at: z.string()
});

export const workflowSummarySchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  review_id: z.string(),
  review_title: z.string(),
  project_id: z.string().nullable(),
  project_title: z.string(),
  mode: z.string(),
  state: z.string(),
  created_at: z.string(),
  selected_agents: z.array(z.string()),
  top_risks: z.array(z.string()),
  finding_count: z.number(),
  has_report: z.boolean()
});

export const providerConnectionSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  adapter: z.string(),
  name: z.string(),
  config: z.record(z.string(), z.unknown()),
  has_credentials: z.boolean()
});

export const modelRecordSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  provider_connection_id: z.string(),
  model_identifier: z.string(),
  capabilities: z.array(z.string()),
  provenance: z.string(),
  verified: z.boolean(),
  probe_result: z.record(z.string(), z.unknown()).default({})
});

export const catalogueModelSchema = z.object({
  model_identifier: z.string(),
  capabilities: z.array(z.string()).default([]),
  provenance: z.string().optional(),
  verified: z.boolean().optional(),
  probe_result: z.record(z.string(), z.unknown()).default({})
});

export const modelProfileSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  name: z.string(),
  agent_key: z.string(),
  model_record_id: z.string(),
  explicit_pin: z.boolean()
});

export const evaluationResultSchema = z.object({
  workspace_id: z.string(),
  fixture_count: z.number(),
  metrics: z.record(z.string(), z.number()),
  adversarial_fixtures: z.array(z.string()),
  live_smoke_tests: z.string()
});

export const governanceSchema = z.object({
  workspace_id: z.string(),
  provider_allowlist: z.array(z.string()),
  model_allowlist: z.array(z.string()),
  data_classification_allowlist: z.array(z.string()),
  region_allowlist: z.array(z.string()),
  purpose_allowlist: z.array(z.string()),
  approved_domains: z.array(z.string()),
  retention_days: z.number(),
  preserve_historical_reports: z.boolean(),
  legal_hold: z.boolean(),
  mfa_required: z.boolean(),
  sso_provider: z.string().nullable(),
  custom_branding: z.record(z.string(), z.unknown()),
  updated_at: z.string()
});

export const enterpriseMemberSchema = z.object({
  workspace_id: z.string(),
  user_id: z.string(),
  email: z.string(),
  role: z.string()
});

export const enterpriseAuditSchema = z.object({
  id: z.string(),
  workspace_id: z.string().nullable(),
  actor_user_id: z.string().nullable(),
  action: z.string(),
  metadata: z.record(z.string(), z.unknown()),
  created_at: z.string()
});

export const enterpriseNotificationSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  user_id: z.string().nullable(),
  kind: z.string(),
  title: z.string(),
  body: z.string(),
  read: z.boolean(),
  created_at: z.string()
});

export const enterpriseOperationsSchema = z.object({
  run_volume: z.number(),
  failure_rate: z.number(),
  security_events: z.number(),
  queue_depth: z.number(),
  tracing_redaction: z.string(),
  quotas: z.record(z.string(), z.number()),
  backup_restore: z.record(z.string(), z.number())
});

export const modelComparisonSchema = z.object({
  workspace_id: z.string(),
  models: z.array(z.object({
    model_identifier: z.string(),
    quality: z.number(),
    cost: z.number(),
    latency_ms: z.number(),
    failure_rate: z.number(),
    capability_coverage: z.number()
  }))
});

export const apiTokenSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  name: z.string(),
  token_prefix: z.string(),
  scopes: z.array(z.string()),
  rate_limit_per_minute: z.number(),
  revoked: z.boolean(),
  plain_token: z.string().nullable().optional()
});

export const webhookSchema = z.object({
  id: z.string(),
  workspace_id: z.string(),
  name: z.string(),
  url: z.string(),
  events: z.array(z.string()),
  enabled: z.boolean(),
  signing_secret: z.string().nullable().optional()
});
